File: code/ch04/4.2.2.24game.py
from typing import List


class Solution:
    def judgePoint24(self, nums: List[int]) -> bool:
        return self.compute([float(i) for i in nums], 4)

    def compute(self, nums: List[int], n: int) -> bool:
        if n == 1:
            return abs(nums[0] - 24) < 0.000001
        new_nums = [0] * 4

        for left in range(n - 1):
            for right in range(left + 1, n):
                index = 0
                for i in range(n):
                    if i != left and i != right:
                        new_nums[index] = nums[i]
                        index += 1

                new_nums[index] = nums[left] + nums[right]
                if self.compute(new_nums, index + 1):
                    return True
                new_nums[index] = nums[left] - nums[right]
                if self.compute(new_nums, index + 1):
                    return True
                new_nums[index] = nums[right] - nums[left]
                if self.compute(new_nums, index + 1):
                    return True
                new_nums[index] = nums[right] * nums[left]
                if self.compute(new_nums, index + 1):
                    return True
                if nums[left] != 0:
                    new_nums[index] = nums[right] / nums[left]
                    if self.compute(new_nums, index + 1):
                        return True
                if nums[right] != 0:
                    new_nums[index] = nums[left] / nums[right]
                    if self.compute(new_nums, index + 1):
                        return True
        return False

File: code/ch04/4.2.2.24game.cpp
class Solution {
private:
    bool compute(vector<double>& nums) {
        if (nums.size() == 1)
            return fabs(nums[0] - 24.0) < 1e-8;
        
        for (int left = 0; left < nums.size() - 1; left++) {
            for (int right = left + 1; right < nums.size(); right++) {
                vector<double> tmp;
                for (int k = 0; k < nums.size(); k++)
                    if (k != left && k != right) tmp.push_back(nums[k]);
                
                tmp.push_back(nums[left] + nums[right]);
                if (compute(tmp))
                    return  true;
                
                tmp.back() = nums[left] * nums[right];
                if (compute(tmp))
                    return true;
                
                tmp.back() = nums[left] - nums[right];
                if (compute(tmp))
                    return true;
                
                tmp.back() = nums[left] / nums[right];
                if (compute(tmp))
                    return true;
                
                tmp.back() = nums[right] - nums[left];
                if (compute(tmp))
                    return true;
                
                tmp.back() = nums[right] / nums[left];
                if (compute(tmp))
                    return true;
            }
        }
        
        return false;
    }
public:
    bool judgePoint24(vector<int>& nums) {
        vector<double> tmp(nums.begin(), nums.end());
        return compute(tmp);
    }
};
